src/assignment1.js

```diff
@@ -1,5 +1,115 @@
+    const figure = {
+        color: '#000',
+        velocity: [2, 2],
+        changeDirection: function (index) {
+            this.velocity[index] = -this.velocity[index];
+        },
+        checkIfContainsAndChangeDirection: function (width, height) {
+            if (!this.containsX(width)) {
+                this.changeDirection(0);
+            }
+            if (!this.containsY(height)) {
+                this.changeDirection(1);
+            }
+        },
+    };
+    string.split('/').forEach((value) => {
+        const pair = value.split(':');
+        figure[pair[0]] = pair[0] === 'shape' || pair[0] === 'color' ? pair[1] : pair[1].includes(',') ?
+            pair[1].split(',').map((stringValue) => parseInt(stringValue)) : parseInt(pair[1]);
+    });
+    if (!validShapes.includes(figure.shape) || figure.center === undefined) {
+        return false;
+    }
+    addSpecifics(figure);
+    return figure;
+}
+
+function addSpecifics(figure) {
+    switch (figure.shape) {
+        case 'triangle':
+            if (figure.length === undefined) {
+                figure.length = 10;
+            }
+            figure.draw = function (context) {
+                context.fillStyle = this.color;
+                const height = sin60 * this.length;
+                context.beginPath();
+                context.moveTo(this.center[0], this.center[1] - 2 * height / 3);
+                context.lineTo(this.center[0] + this.length / 2, this.center[1] + height / 3);
+                context.lineTo(this.center[0] - this.length / 2, this.center[1] + height / 3);
+                context.closePath();
+                context.fill();
+            };
+            figure.containsX = function (width) {
+                return !(this.center[0] - this.length / 2 <= 0 || this.center[0] + this.length / 2 >= width);
+            };
+            figure.containsY = function (height) {
+                return !(this.center[1] - 2 * sin60 * this.length / 3 <= 0 || this.center[1] + sin60 * this.length / 3 >= height);
+            };
+            break;
+        case 'rect':
+            if (figure.width === undefined) {
+                figure.width = 10;
+            }
+            if (figure.height === undefined) {
+                figure.height = 10;
+            }
+            figure.draw = function (context) {
+                context.fillStyle = this.color;
+                context.fillRect(this.center[0] - (this.width / 2), this.center[1] - (this.height / 2), this.width, this.height);
+            };
+            figure.containsX = function (width) {
+                return !(this.center[0] - this.width / 2 <= 0 || this.center[0] + this.width / 2 >= width);
+            };
+            figure.containsY = function (height) {
+                return !(this.center[1] - this.height / 2 <= 0 || this.center[1] + this.height / 2 >= height);
+            };
+            break;
+        case 'circle':
+            if (figure.radius === undefined) {
```

Comment:
No need to check if value is equal to undefined, `if (figure.radius)` is enough, since all non-negative values are truthy and even if radius is set to 0, which is falsey, it obviously should get it's default value.

Reply:
done